stories/mse-eme-player.tsx

```diff
@@ -0,0 +1,261 @@
+
+const mimeType = 'audio/mp4; codecs="mp4a.40.2"'
+
+const { KEY, KID } = {
+  KEY: '7f412f0575f44f718259beef56ec7771',
+  KID: '2fef8ad812df429783e9bf6e5e493e53',
+}
+
+const config = [
+  {
+    initDataTypes: ['cenc'], // keyids, cenc
+    audioCapabilities: [{ contentType: mimeType }],
+  },
+]
+
+let pendingPromise
+
+function EnsureMediaKeysCreated(el, keySystem, options) {
+  if (pendingPromise) {
+    return pendingPromise
+  }
+
+  pendingPromise = navigator
+    .requestMediaKeySystemAccess(keySystem, options)
+    .then((keySystemAccess) => keySystemAccess.createMediaKeys())
+    .then((mediaKeys) => el.setMediaKeys(mediaKeys))
+
+  return pendingPromise
+}
+
+function ArrayBufferToString(arr) {
+  let str = ''
+  const view = new Uint8Array(arr)
+  for (let i = 0; i < view.length; i++) {
+    str += String.fromCharCode(view[i])
+  }
+  return str
+}
+
+function StringToArrayBuffer(str) {
+  const arr = new ArrayBuffer(str.length)
+  const view = new Uint8Array(arr)
+  for (let i = 0; i < str.length; i++) {
+    view[i] = str.charCodeAt(i)
+  }
+  return arr
+}
+
+function HexToBase64(hex) {
+  let bin = ''
+  for (let i = 0; i < hex.length; i += 2) {
+    bin += String.fromCharCode(parseInt(hex.substr(i, 2), 16))
+  }
+  return window.btoa(bin).replace(/=/g, '').replace(/\+/g, '-').replace(/\//g, '_')
+}
+
+function handleMessage(ev) {
+  const msgStr = JSON.parse(ArrayBufferToString(ev.message))
+
+  const outKeys = []
+
+  outKeys.push({
+    kty: 'oct',
+    alg: 'A128KW',
+    kid: HexToBase64(KID),
+    k: HexToBase64(KEY),
+  })
+
+  const update = JSON.stringify({
+    keys: outKeys,
+    type: msgStr.type,
+  })
+
+  ev.target.update(StringToArrayBuffer(update))
+}
+
+interface ChunkData {
+  name: string
+  data: Array<number> | null
+  segmentTime: number
+  chunkIndex: number
+}
+
+class MediaSourcePlayer extends PureComponent<unknown, { volumeText: string }> {
+  player = createRef<AudioPlayer>()
+  mediaSource = null
+  loadingChunk = false
+  chunks = []
+  whenBufferUpdateEndCallbacks = []
+
+  state = {
+    audioSrc: '',
+    srcDuration: void 0,
+  }
+
+  onEncrypted(ev: MediaEncryptedEvent): void {
+    const audio = ev.target
+    EnsureMediaKeysCreated(audio, 'org.w3.clearkey', config).then(() => {
+      audio.sessions = []
+
+      const session = audio.mediaKeys.createSession()
+      audio.sessions.push(session)
+      session.addEventListener('message', handleMessage)
+
+      return session.generateRequest(ev.initDataType, ev.initData)
+    })
+
+    // and for the storybook action
+    this.props.onEncrypted(ev)
+  }
+
+  onSeek(audio: HTMLAudioElement, time: number): void {
+    const playlist = this.chunks
+    const bufferIsCompleteUpToNewCurrentTime = playlist.every((chunk, i) => {
+      const chunkStartTime = chunk.segmentTime * i
+      return chunkStartTime > time || chunk.data
+    })
+
+    if (bufferIsCompleteUpToNewCurrentTime) {
```

Comment:
It's a hard decision for whether putting these SME/EME logic in the consumer app or inside the library.

I mentioned I'd like the changes to the library to be minimal, but I doubt that if it's really good.
In your personal opinion, do you suggest to put the most MSE EME logic in consumer side or inside the library?

Reply:
Good point. Adding this to the library will increase complexity but make life easier to most use cases. This scenario is when all chunks have the same duration, otherwise we need a manifest (like `.mpd` in DASH or `.m3u8` in HLS) and that logic can be different. We could put it in the library with 2 ideas:
 - expose a function for this logic so the consumer can override this and run 100% their logic. The default would be the library logic
 - we could pass the MSE/EME related props inside a object like `useMSE: {onSeek, srcDuration.... etc}` or just `eme`. Like this the library can more easy separate the "mode" in which it is working and organise the code more easy.

Let me know your thoughts

Reply:
I don't have any MSE/EME dev experience, do you think "all chunks have the same duration" is a common scenario? If so, we should definitely put the logic inside the logic and provide props for overwriting; if not, we can leave it as what you have implemented, and provided examples in storybook.

Reply:
I don't think all chunks with same duration is common. The last chunk is very likely custom since it's the left over of a x seconds division. Also the first chunk can be smaller so the user gets some sound and the app can continue with chunk nr.2 while playing the first. 

Would it be ok for you to:
 - `a)` pass all MSE props inside a props object like: `useMSE: {onSeek, srcDuration, etc}`?
 - `b)` put the chunk logic inside the library and for that pass the chunk array to the library as a new `audioChunks` prop so the player can do chunk logic?

I think this ☝️could be a good solution API. We offer some heavy logic fixed inside the library but give the tools to have custom logic.